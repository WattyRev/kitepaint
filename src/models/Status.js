import PropTypes from "prop-types";

/**
 * Status can be applied to any model to indicate it's public visibility.
 */
export const statusProp = PropTypes.oneOf(["0", "1", "2"]);
const Status = {
  "0": "Private",
  "1": "Unlisted",
  "2": "Public",
  PRIVATE: "0",
  UNLISTED: "1",
  PUBLIC: "2"
};
export default Status;
