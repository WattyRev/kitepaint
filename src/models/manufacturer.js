import PropTypes from "prop-types";

/**
 * A manufacturer is a company that manufactures kites and owns one or more products.
 */
const manufacturerState = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  logo: PropTypes.string.isRequired,
  website: PropTypes.string
});

export default manufacturerState;

/**
 * A mock manufacturer used for testing
 * @type {Object}
 */
const getMockManufacturer = () => ({
  id: "abc",
  name: "Krazy Kites",
  logo: "abc.jpg",
  website: "http://krazykites.com"
});
export { getMockManufacturer };
