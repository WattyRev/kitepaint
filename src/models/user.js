import PropTypes from "prop-types";

/**
 * A user is an individual user account. It contains limited details that are available publically.
 */
const userShape = PropTypes.shape({
  username: PropTypes.string.isRequired,
  loginid: PropTypes.string.isRequired
});

export default userShape;

/**
 * A mock user used for testing.
 */
const getMockUser = () => ({
  username: "FlyFucker",
  loginid: "abc"
});

export { getMockUser };
