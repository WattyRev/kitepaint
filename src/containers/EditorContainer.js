import React from "react";
import PropTypes from "prop-types";
import { fromJS } from "immutable";
import { connect } from "react-redux";
import { CREATE_DESIGN, UPDATE_DESIGN } from "../redux/actions";
import designShape from "../models/design";
import productShape from "../models/product";
import Status from "../models/status";
import { isEmbedded, defaultBackground } from "../constants/embed";
import ErrorPage from "../components/ErrorPage";
import { softCompareStrings, makeCancelable, embedAllowed } from "../utils";

const appliedColorsShape = PropTypes.objectOf(
  PropTypes.shape({
    name: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired
  })
);

const productAppliedColorsShape = PropTypes.objectOf(appliedColorsShape);

export { appliedColorsShape, productAppliedColorsShape };

/**
 * Manages the overall state of the editor. And provides a means to save changes.
 */
export class EditorContainer extends React.Component {
  static propTypes = {
    /**
     * The product being edited
     */
    product: productShape.isRequired,
    /**
     * The default color (name) to be selected. Will default to the first color on the product
     * otherwise.
     */
    defaultColor: PropTypes.string,
    /**
     * An existing design being edited.
     */
    design: designShape,
    /**
     * The default variation (name) to be selected. Will default to the first variation on the
     * product otherwise.
     */
    defaultVariation: PropTypes.string,
    /**
     * A function that renders content
     */
    children: PropTypes.func.isRequired,
    /**
     * A function called when we want to save the design. Provided by Redux.
     */
    onSave: PropTypes.func.isRequired,
    /**
     * A function called when we want to update a design. Provided by Redux.
     */
    onUpdate: PropTypes.func.isRequired
  };

  /**
   * Parses the variations from the provided design in order to determine what colors have been
   * applied.
   * @return {Object}
   */
  _generateAppliedColors() {
    if (!this.props.design) {
      return {};
    }
    const colors = this.props.product.colors;
    return this.props.design.variations.reduce((accumulated, variation) => {
      const { svg, name } = variation;
      const render = new window.DOMParser().parseFromString(svg, "text/xml");
      const panels = render.querySelectorAll("[data-id]");
      const appliedColors = {};
      for (let i = 0; i < panels.length; i++) {
        const panel = panels[i];
        const color = panel.getAttribute("fill");
        if (color) {
          const colorMatch = colors.find(storedColor =>
            softCompareStrings(storedColor.color, color)
          );
          const colorName = colorMatch ? colorMatch.name : color;
          appliedColors[panel.getAttribute("data-id")] = {
            color,
            name: colorName
          };
        }
      }

      accumulated[name] = appliedColors;
      return accumulated;
    }, {});
  }

  constructor(props, ...rest) {
    super(props, ...rest);

    // Use the first color, or the one specified
    let currentColor = props.product.colors[0];
    if (props.defaultColor) {
      currentColor = props.product.colors.find(color =>
        softCompareStrings(color.name, props.defaultColor)
      );
    }

    // Use the first variation or the one specified
    let currentVariation = props.product.variations[0];
    if (props.defaultVariation) {
      currentVariation = props.product.variations.find(variation =>
        softCompareStrings(variation.name, props.defaultVariation)
      );
    }

    this.state = {
      background: defaultBackground || null,
      hideOutlines: false,
      /**
       * The currently selected color.
       * @type {Object}
       */
      currentColor,
      /**
       * The currently selected variation.
       * @type {Object}
       */
      currentVariation,
      /**
       * The currently applied colors.
       * Example
       * {
       *   "my-variation": {
       *     p1: {
       *       name: "black",
       *       color: "#000"
       *     }
       *   }
       * }
       * @type {Object}
       */
      appliedColors: this._generateAppliedColors()
    };
  }

  /*
   * Cancels any pending promises before being unmounted.
   */
  componentWillUnmount() {
    this.cancelablePromises.forEach(promise => promise.cancel());
  }

  /**
   * An array of promises that may need to be cancelled when the component is unmounted
   */
  cancelablePromises = [];

  /**
   * Handles when a different color is selected by updating state.
   * @param  {String} colorName The name of the newly selected color
   */
  handleColorSelection = colorName => {
    const currentColor = this.props.product.colors.find(color =>
      softCompareStrings(color.name, colorName)
    );
    this.setState({
      currentColor
    });
  };

  /**
   * Handles when a different variation is selected by updating state.
   * @param  {String} variationName The name of the newly selected variation
   */
  handleVariationSelection = variationName => {
    const currentVariation = this.props.product.variations.find(variation =>
      softCompareStrings(variation.name, variationName)
    );
    this.setState({
      currentVariation
    });
  };

  /**
   * Set the color for the specified id on the current variation to the current color.
   * @param  {String} id The ID of the panel, taken from data-id on the element.
   */
  handleColorApplied = id => {
    const appliedColorsMap = fromJS(this.state.appliedColors);
    const appliedColors = appliedColorsMap
      .setIn([this.state.currentVariation.name, id], this.state.currentColor)
      .toJS();
    this.setState({
      appliedColors
    });
  };

  /**
   * Generates the design variations based on the product variations and the applied colors.
   * @return {Object[]} Each object contains name, primary, and svg.
   */
  generateDesignVariations = () => {
    const appliedColors = this.state.appliedColors;
    const productVariations = this.props.product.variations;

    // Build each variation
    return productVariations.map((variation, index) => {
      // Render the blank variation from the product in memory
      const render = new window.DOMParser().parseFromString(
        variation.svg,
        "text/xml"
      );
      const colorMap = appliedColors[variation.name] || {};

      // Apply each color to the rendered variation
      Object.keys(colorMap).forEach(id => {
        const color = colorMap[id].color;
        const panel = render.querySelector(`[data-id="${id}"]`);
        panel.setAttribute("fill", color);
      });

      // Get the new SVG string from the render.
      const svg = render.querySelector("svg").outerHTML;

      // Return the variation
      return {
        name: variation.name,
        primary: !index,
        svg
      };
    });
  };

  /**
   * Handles save by parsing data and submitting a request to create a new design. Redirects to that
   * design's edit page when successful.
   * @param  {Object} data must contain name and user(id)
   */
  handleSave = data => {
    const { name, user } = data;
    const design = {
      name,
      user,
      product: this.props.product.id,
      variations: this.generateDesignVariations(),
      status: user === "0" ? Status.PUBLIC : Status.UNLISTED
    };
    const promise = makeCancelable(this.props.onSave(design));
    promise.promise.then(response => {
      const designId = response.data.id;
      if (data.user === "0") {
        // If the design was created anonymously, go to the view page
        window.location.replace(`/view/${designId}`);
      } else {
        window.location.replace(`/edit/${designId}`);
      }
    });
    this.cancelablePromises.push(promise);
  };

  /**
   * Handles update by parsing data and submitting a request to update the design.
   */
  handleUpdate = () => {
    const design = {
      id: this.props.design.id,
      variations: this.generateDesignVariations()
    };
    this.props.onUpdate(design);
  };

  /**
   * Gets the applied colors for the current variation
   */
  getCurrentVariationColors = () => {
    const currentVariationName = this.state.currentVariation.name;
    return this.state.appliedColors[currentVariationName] || {};
  };

  /**
   * Handles auto fill by merging the colors applied to the current variation on top of the colors
   * applied to every other variation.
   */
  handleAutofill = () => {
    const currentColors = this.getCurrentVariationColors();
    const previousAppliedColors = this.state.appliedColors;
    const variations = this.props.product.variations;
    const appliedColors = variations.reduce((accumulated, variation) => {
      accumulated[variation.name] = {
        ...(previousAppliedColors[variation.name] || {}),
        ...currentColors
      };
      return accumulated;
    }, {});
    this.setState({ appliedColors });
  };

  handleChangeBackground = value => this.setState({ background: value });

  handleToggleHideOutlines = () =>
    this.setState({ hideOutlines: !this.state.hideOutlines });

  render() {
    if (
      isEmbedded &&
      this.props.product &&
      !embedAllowed(this.props.product.embed.split(","))
    ) {
      return (
        <ErrorPage
          errorCode={401}
          errorMessage="Embedding of this page is not permitted."
        />
      );
    }
    const data = {
      actions: {
        applyColor: this.handleColorApplied,
        autofill: this.handleAutofill,
        changeBackground: this.handleChangeBackground,
        toggleHideOutlines: this.handleToggleHideOutlines,
        save: this.handleSave,
        selectColor: this.handleColorSelection,
        selectVariation: this.handleVariationSelection,
        update: this.handleUpdate
      },
      props: {
        appliedColors: this.state.appliedColors,
        background: this.state.background,
        hideOutlines: this.state.hideOutlines,
        currentColor: this.state.currentColor,
        currentVariation: this.state.currentVariation,
        currentVariationColors: this.getCurrentVariationColors()
      }
    };
    return this.props.children(data);
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  onSave: CREATE_DESIGN,
  onUpdate: UPDATE_DESIGN
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditorContainer);
