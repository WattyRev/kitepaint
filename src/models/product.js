import PropTypes from "prop-types";
import { getMockManufacturer } from "./manufacturer";
import Status, { statusProp } from "./status";

/**
 * A product is a kite that can be customized to create a design.
 */
const productShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  )
});

export default productShape;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = () => ({
  id: "product-1",
  embed: "krazykites.com",
  name: "Krazy Kite",
  manufacturer: getMockManufacturer().id,
  url: "http://krazykites.com/krazy-kite",
  colors: [
    {
      name: "red",
      color: "#ff0000"
    },
    {
      name: "black",
      color: "#000000"
    }
  ],
  variations: [
    {
      name: "Standard",
      svg: "<div>Kool Kite</div>"
    }
  ],
  status: Status.PUBLIC
});

export { getMockProduct };
