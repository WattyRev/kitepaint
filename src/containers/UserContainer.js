import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getUser, getUserRecognition } from "../redux/modules/user";
import { SET_RECOGNIZED_USER, LOG_IN, LOG_OUT } from "../redux/actions";

/**
 * Maintains a local storage variable that indicates if the user is recognized as a registered user.
 * TODO this needs to be done in redux, because there will need to be more than one of these, and
 * they need to stay in sync.
 *
 * @param {Function} children
 */
class UserContainer extends React.Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    onSetRecognition: PropTypes.func.isRequired,
    onLogIn: PropTypes.func.isRequired
  };

  /**
   * Toggles the variable in state and in local storage
   */
  toggleRecognition = () => {
    const { isRecognizedUser } = this.props;
    this.props.onSetRecognition(!isRecognizedUser);
  };

  render() {
    const {
      firstName,
      id,
      isLoggedIn,
      isLoggingIn,
      lastName,
      username
    } = this.props.user;
    return this.props.children({
      actions: {
        toggleRecognition: this.toggleRecognition,
        logIn: this.props.onLogIn,
        logOut: this.props.onLogOut
      },
      props: {
        firstName,
        id,
        isLoggedIn,
        isLoggingIn,
        isRecognizedUser: this.props.isRecognizedUser,
        lastName,
        username
      }
    });
  }
}

const mapStateToProps = state => ({
  user: getUser(state),
  isRecognizedUser: getUserRecognition(state)
});

const mapDispatchToProps = {
  onLogIn: LOG_IN,
  onLogOut: LOG_OUT,
  onSetRecognition: SET_RECOGNIZED_USER
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserContainer);
