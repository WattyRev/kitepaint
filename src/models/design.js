import PropTypes from "prop-types";
import { getMockProduct } from "./product";
import Status, { statusProp } from "./status";

/**
 * A design is a user-created entity. It represetns a saved design of a particular product.
 */
const designShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  // A string representing the created date
  created: PropTypes.string.isRequired,
  // A string representing the updated date
  updated: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  // The ID of the cooresponding product
  product: PropTypes.string.isRequired,
  // The ID of the user that created the design
  user: PropTypes.string,
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired,
      primary: PropTypes.bool.isRequired
    })
  ),
  // The status of the cooresponding product
  productStatus: statusProp.isRequired,
  // The status of the design
  status: statusProp.isRequired
});

export default designShape;

/**
 * A mock design used for testing
 * @type {Object}
 */
const getMockDesign = () => ({
  id: "design-1",
  created: "01/01/2018",
  updated: "01/02/2018",
  name: "picasso",
  product: getMockProduct().id,
  variations: [
    {
      name: "Standard",
      svg: "<div>picasso standard</div>",
      primary: true
    }
  ],
  status: Status.PRIVATE,
  productStatus: Status.PUBLIC
});
export { getMockDesign };
