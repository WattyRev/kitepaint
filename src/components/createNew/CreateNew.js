import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import ProductContainer from "../../containers/ProductContainer";
import EditorContainer from "../../containers/EditorContainer";
import UserContainer from "../../containers/UserContainer";
import { Text, H3, PageLoader } from "../../theme";
import Toolbar from "../editor/Toolbar";
import Sidebar from "../editor/Sidebar";
import Canvas from "../editor/Canvas";

const PageLayout = styled.div`
  display: flex;
  justify-content: stretch;
  position: relative;

  .product-notes {
    position: absolute;
    background: rgba(0, 0, 0, 0.8);
    bottom: 0;
    right: 0;
    padding: 8px;
    max-width: calc(100% - 200px);
  }
`;

/**
 * A coordinating component that builds the CreatNew page.
 * The CreateNew page is provided a productId from the router, and is intended to create a new
 * design based on that product.
 */
const CreateNew = ({ match }) => (
  <ProductContainer productId={match.params.productId}>
    {productData =>
      productData.props.isLoading ? (
        <PageLoader />
      ) : (
        <EditorContainer product={productData.props.product}>
          {editorData => (
            <React.Fragment>
              <UserContainer>
                {userData => (
                  <Toolbar
                    onSave={
                      userData.props.isLoggedIn
                        ? name =>
                            editorData.actions.save({
                              name,
                              user: userData.props.id
                            })
                        : null
                    }
                    onShare={() => {}}
                    onAutofill={editorData.actions.autofill}
                    onReset={() => {}}
                    onHideOutlines={() => {}}
                    onBackgroundChange={() => {}}
                  />
                )}
              </UserContainer>
              <PageLayout>
                <Sidebar
                  product={productData.props.product}
                  manufacturer={productData.props.manufacturer}
                  selectedVariation={editorData.props.currentVariation.name}
                  selectedColor={editorData.props.currentColor.name}
                  onColorSelect={editorData.actions.selectColor}
                  onVariationSelect={editorData.actions.selectVariation}
                  appliedColors={editorData.props.appliedColors}
                />
                <Canvas
                  colorMap={editorData.props.currentVariationColors}
                  svg={editorData.props.currentVariation.svg}
                  onClick={editorData.actions.applyColor}
                  currentColor={editorData.props.currentColor.name}
                />
                {productData.props.product.notes &&
                  !!productData.props.product.notes.length && (
                    <div className="product-notes">
                      <H3 isLight>Notes:</H3>
                      <ul>
                        {productData.props.product.notes.map(
                          (note, index) =>
                            note.trim() && (
                              <Text isLight as="li" key={note + index}>
                                {note}
                              </Text>
                            )
                        )}
                      </ul>
                    </div>
                  )}
              </PageLayout>
            </React.Fragment>
          )}
        </EditorContainer>
      )
    }
  </ProductContainer>
);

CreateNew.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      productId: PropTypes.string.isRequired
    })
  })
};

export default CreateNew;
