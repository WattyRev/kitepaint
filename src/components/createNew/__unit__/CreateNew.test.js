import React from "react";
import { shallow } from "enzyme";
import CreateNewContainer from "../../../containers/CreateNewContainer";
import EditorContainer from "../../../containers/EditorContainer";
import UserContainer from "../../../containers/UserContainer";
import { getMockProduct } from "../../../models/product";
import { getMockManufacturer } from "../../../models/manufacturer";
import CreateNew from "../CreateNew";

describe("CreateNew", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      match: {
        params: {
          productId: "abc"
        }
      }
    };
  });
  it("renders", () => {
    expect.assertions(4);
    const wrapper = shallow(<CreateNew {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();

    // Drill in to product container
    const productData = {
      props: {
        product: getMockProduct(),
        manufacturer: getMockManufacturer()
      }
    };
    const CreateNewContainerWrapper = shallow(
      <div>
        {wrapper.find(CreateNewContainer).prop("children")(productData)}
      </div>
    );
    expect(CreateNewContainerWrapper).toMatchSnapshot();

    // Drill in to editor container
    const editorData = {
      actions: {
        selectVariation: jest.fn(),
        selectColor: jest.fn(),
        applyColor: jest.fn(),
        save: jest.fn()
      },
      props: {
        currentColor: {
          name: "blue"
        },
        currentVariation: {
          name: "Standard",
          svg: ""
        },
        appliedColors: {},
        currentVariationColors: {}
      }
    };
    const editorContainerWrapper = shallow(
      <div>
        {CreateNewContainerWrapper.find(EditorContainer).prop("children")(
          editorData
        )}
      </div>
    );
    expect(editorContainerWrapper).toMatchSnapshot();

    // Drill in to user container
    const userData = {
      props: {
        isLoggedIn: true,
        id: "123"
      }
    };
    const userContainerWrapper = shallow(
      <div>
        {editorContainerWrapper.find(UserContainer).prop("children")(userData)}
      </div>
    );
    expect(userContainerWrapper).toMatchSnapshot();
  });
});
