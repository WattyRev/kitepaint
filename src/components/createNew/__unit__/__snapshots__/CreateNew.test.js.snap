// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNew renders 1`] = `
<Connect(CreateNewContainer)
  productId="abc"
>
  <Component />
</Connect(CreateNewContainer)>
`;

exports[`CreateNew renders 2`] = `
<div>
  <Connect(EditorContainer)
    product={
      Object {
        "colors": Array [
          Object {
            "color": "#ff0000",
            "name": "red",
          },
          Object {
            "color": "#000000",
            "name": "black",
          },
        ],
        "id": "product-1",
        "manufacturer": "manufacturer-1",
        "name": "Krazy Kite",
        "status": "2",
        "url": "http://krazykites.com/krazy-kite",
        "variations": Array [
          Object {
            "name": "Standard",
            "svg": "<div>Kool Kite</div>",
          },
        ],
      }
    }
  >
    <Component />
  </Connect(EditorContainer)>
</div>
`;

exports[`CreateNew renders 3`] = `
<div>
  <Connect(UserContainer)>
    <Component />
  </Connect(UserContainer)>
  <styled.div>
    <Sidebar
      appliedColors={Object {}}
      manufacturer={
        Object {
          "id": "manufacturer-1",
          "logo": "abc.jpg",
          "name": "Krazy Kites",
          "website": "http://krazykites.com",
        }
      }
      onColorSelect={[MockFunction]}
      onVariationSelect={[MockFunction]}
      product={
        Object {
          "colors": Array [
            Object {
              "color": "#ff0000",
              "name": "red",
            },
            Object {
              "color": "#000000",
              "name": "black",
            },
          ],
          "id": "product-1",
          "manufacturer": "manufacturer-1",
          "name": "Krazy Kite",
          "status": "2",
          "url": "http://krazykites.com/krazy-kite",
          "variations": Array [
            Object {
              "name": "Standard",
              "svg": "<div>Kool Kite</div>",
            },
          ],
        }
      }
      selectedColor="blue"
      selectedVariation="Standard"
    />
    <Canvas
      colorMap={Object {}}
      currentColor="blue"
      onClick={[MockFunction]}
      svg=""
    />
  </styled.div>
</div>
`;

exports[`CreateNew renders 4`] = `
<div>
  <Toolbar
    onBackgroundChange={[Function]}
    onHideOutlines={[Function]}
    onReset={[Function]}
    onSave={[Function]}
    onShare={[Function]}
  />
</div>
`;
